File: packages/yakr-core/src/yakr_core/ratchet.py
from __future__ import annotations

import struct
from dataclasses import dataclass, field

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import x25519

from yakr_core.crypto import hkdf_derive, x25519_shared_secret, xchacha_decrypt, xchacha_encrypt

RATCHET_MAGIC = b"YKDR2"
MAX_SKIP_GAP = 128
MAX_SKIPPED_KEYS = 256
ROOT_INFO = b"yakr/v1.0/double-ratchet-root"
RK_INFO = b"yakr/v1.0/double-ratchet-rk"
CK_INFO = b"yakr/v1.0/double-ratchet-ck"
SEND_CHAIN_INFO = b"yakr/v1.0/double-ratchet-send"
RECV_CHAIN_INFO = b"yakr/v1.0/double-ratchet-recv"


def _kdf_rk(root_key: bytes, dh_output: bytes) -> tuple[bytes, bytes]:
    material = hkdf_derive(root_key, RK_INFO, salt=dh_output, length=64)
    return material[:32], material[32:]


def _kdf_ck(chain_key: bytes) -> tuple[bytes, bytes]:
    material = hkdf_derive(chain_key, CK_INFO, length=64)
    return material[:32], material[32:]


def _pub_bytes(key: x25519.X25519PrivateKey) -> bytes:
    return key.public_key().public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )


def _load_private(raw: bytes) -> x25519.X25519PrivateKey:
    return x25519.X25519PrivateKey.from_private_bytes(raw)


@dataclass
class RatchetState:
    """X25519 double ratchet with symmetric send/recv chains."""

    root_key: bytes
    dh_self_private: bytes
    dh_self_public: bytes
    dh_peer_public: bytes | None = None
    send_chain_key: bytes | None = None
    recv_chain_key: bytes | None = None
    send_n: int = 0
    recv_n: int = 0
    prev_send_n: int = 0
    skipped_keys: dict[str, str] = field(default_factory=dict)
    hybrid: bool = False
    pending_pairing_dh_ratchet_peer: bytes | None = None

    @classmethod
    def from_master(
        cls,
        master_secret: bytes,
        *,
        is_initiator: bool,
        hybrid: bool = False,
        ratchet_private: x25519.X25519PrivateKey | None = None,
    ) -> RatchetState:
        root_key = hkdf_derive(master_secret, ROOT_INFO)
        send_chain = hkdf_derive(root_key, SEND_CHAIN_INFO)
        recv_chain = hkdf_derive(root_key, RECV_CHAIN_INFO)
        if not is_initiator:
            send_chain, recv_chain = recv_chain, send_chain
        if ratchet_private is None:
            ratchet_private = x25519.X25519PrivateKey.generate()
        dh_public = _pub_bytes(ratchet_private)
        return cls(
            root_key=root_key,
            dh_self_private=ratchet_private.private_bytes(
                encoding=serialization.Encoding.Raw,
                format=serialization.PrivateFormat.Raw,
                encryption_algorithm=serialization.NoEncryption(),
            ),
            dh_self_public=dh_public,
            send_chain_key=send_chain,
            recv_chain_key=recv_chain,
            hybrid=hybrid,
        )

    def to_dict(self) -> dict[str, str | int | bool | dict[str, str]]:
        import base64

        return {
            "version": 2,
            "root_key": base64.urlsafe_b64encode(self.root_key).decode("ascii").rstrip("="),
            "dh_self_private": base64.urlsafe_b64encode(self.dh_self_private).decode("ascii").rstrip("="),
            "dh_self_public": base64.urlsafe_b64encode(self.dh_self_public).decode("ascii").rstrip("="),
            "dh_peer_public": (
                base64.urlsafe_b64encode(self.dh_peer_public).decode("ascii").rstrip("=")
                if self.dh_peer_public is not None
                else ""
            ),
            "send_chain_key": (
                base64.urlsafe_b64encode(self.send_chain_key).decode("ascii").rstrip("=")
                if self.send_chain_key is not None
                else ""
            ),
            "recv_chain_key": (
                base64.urlsafe_b64encode(self.recv_chain_key).decode("ascii").rstrip("=")
                if self.recv_chain_key is not None
                else ""
            ),
            "send_n": self.send_n,
            "recv_n": self.recv_n,
            "prev_send_n": self.prev_send_n,
            "skipped_keys": self.skipped_keys,
            "hybrid": self.hybrid,
            "pending_pairing_dh_ratchet_peer": (
                base64.urlsafe_b64encode(self.pending_pairing_dh_ratchet_peer).decode("ascii").rstrip("=")
                if self.pending_pairing_dh_ratchet_peer is not None
                else ""
            ),
        }

    @classmethod
    def from_dict(cls, payload: dict[str, str | int | bool | dict[str, str]]) -> RatchetState:
        import base64

        def dec(value: str) -> bytes:
            if not value:
                return b""
            padding = "=" * (-len(value) % 4)
            return base64.urlsafe_b64decode(value + padding)

        if int(payload.get("version", 0)) != 2:
            raise ValueError("unsupported ratchet version; re-pair required")

        peer = dec(str(payload.get("dh_peer_public", "")))
        send_chain = dec(str(payload.get("send_chain_key", "")))
        recv_chain = dec(str(payload.get("recv_chain_key", "")))
        skipped = payload.get("skipped_keys", {})
        if not isinstance(skipped, dict):
            skipped = {}
        pending_peer = dec(str(payload.get("pending_pairing_dh_ratchet_peer", "")))
        return cls(
            root_key=dec(str(payload["root_key"])),
            dh_self_private=dec(str(payload["dh_self_private"])),
            dh_self_public=dec(str(payload["dh_self_public"])),
            dh_peer_public=peer or None,
            send_chain_key=send_chain or None,
            recv_chain_key=recv_chain or None,
            send_n=int(payload.get("send_n", 0)),
            recv_n=int(payload.get("recv_n", 0)),
            prev_send_n=int(payload.get("prev_send_n", 0)),
            skipped_keys={str(k): str(v) for k, v in skipped.items()},
            hybrid=bool(payload.get("hybrid", False)),
            pending_pairing_dh_ratchet_peer=pending_peer or None,
        )

    def _skip_key(self, dh_public: bytes, n: int) -> bytes:
        import base64

        key_id = f"{dh_public.hex()}:{n}"
        stored = self.skipped_keys.get(key_id)
        if stored is None:
            raise KeyError(key_id)
        padding = "=" * (-len(stored) % 4)
        return base64.urlsafe_b64decode(stored + padding)

    def _store_skip(self, dh_public: bytes, n: int, message_key: bytes) -> None:
        import base64

        key_id = f"{dh_public.hex()}:{n}"
        self.skipped_keys[key_id] = base64.urlsafe_b64encode(message_key).decode("ascii").rstrip("=")

    def _pairing_send_init(self, peer_public: bytes) -> None:
        """Initiator first send: derive send chain without recv transition."""
        if len(peer_public) != 32:
            raise ValueError("peer ratchet public key must be 32 bytes")
        self.dh_peer_public = peer_public
        dh_self = _load_private(self.dh_self_private)
        dh_output = x25519_shared_secret(dh_self, peer_public)
        root_key, _recv_unused = _kdf_rk(self.root_key, dh_output)

        new_private = x25519.X25519PrivateKey.generate()
        self.dh_self_private = new_private.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )
        self.dh_self_public = _pub_bytes(new_private)
        dh_output = x25519_shared_secret(new_private, peer_public)
        self.root_key, self.send_chain_key = _kdf_rk(root_key, dh_output)
        self.prev_send_n = self.send_n
        self.send_n = 0

    def _pairing_recv_init(self, peer_public: bytes) -> None:
        """Responder pairing: align recv chain with initiator's post-pairing send chain."""
        if len(peer_public) != 32:
            raise ValueError("peer ratchet public key must be 32 bytes")
        self.dh_peer_public = peer_public
        dh_self = _load_private(self.dh_self_private)
        dh_output = x25519_shared_secret(dh_self, peer_public)
        self.root_key, self.recv_chain_key = _kdf_rk(self.root_key, dh_output)
        self.recv_n = 0

    def _dh_ratchet(self, peer_public: bytes) -> None:
        self.skipped_keys.clear()
        self.dh_peer_public = peer_public
        dh_self = _load_private(self.dh_self_private)
        dh_output = x25519_shared_secret(dh_self, peer_public)
        self.root_key, self.recv_chain_key = _kdf_rk(self.root_key, dh_output)
        self.recv_n = 0

        new_private = x25519.X25519PrivateKey.generate()
        self.dh_self_private = new_private.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )
        self.dh_self_public = _pub_bytes(new_private)
        dh_output = x25519_shared_secret(new_private, peer_public)
        self.root_key, self.send_chain_key = _kdf_rk(self.root_key, dh_output)
        self.prev_send_n = self.send_n
        self.send_n = 0

    def _header_aad(self, *, prev_n: int, message_n: int) -> bytes:
        return RATCHET_MAGIC + self.dh_self_public + struct.pack(">II", prev_n, message_n)

    def encrypt(self, plaintext: bytes) -> bytes:
        if self.pending_pairing_dh_ratchet_peer is not None:
            peer = self.pending_pairing_dh_ratchet_peer
            self.pending_pairing_dh_ratchet_peer = None
            self._pairing_send_init(peer)
        if self.send_chain_key is None:
            raise ValueError("send chain not initialized")
        message_key, self.send_chain_key = _kdf_ck(self.send_chain_key)
        aad = self._header_aad(prev_n=self.prev_send_n, message_n=self.send_n)
        ciphertext = xchacha_encrypt(message_key, plaintext, associated_data=aad)
        header = RATCHET_MAGIC + self.dh_self_public + struct.pack(">II", self.prev_send_n, self.send_n)
        self.send_n += 1
        return header + ciphertext

    def decrypt(self, payload: bytes) -> bytes:
        if len(payload) < len(RATCHET_MAGIC) + 32 + 8:
            raise ValueError("ratchet payload too short")
        if not payload.startswith(RATCHET_MAGIC):
            raise ValueError("invalid ratchet header")
        offset = len(RATCHET_MAGIC)
        peer_public = payload[offset : offset + 32]
        offset += 32
        prev_n, message_n = struct.unpack(">II", payload[offset : offset + 8])
        offset += 8
        ciphertext = payload[offset:]

        if self.dh_peer_public is None:
            self.dh_peer_public = peer_public
        elif self.dh_peer_public != peer_public:
            self._dh_ratchet(peer_public)

        if self.recv_chain_key is None:
            raise ValueError("recv chain not initialized")

        if message_n < self.recv_n:
            try:
                message_key = self._skip_key(peer_public, message_n)
            except KeyError:
                raise ValueError("ratchet message already received") from None
        else:
            gap = message_n - self.recv_n
            if gap > MAX_SKIP_GAP:
                raise ValueError("ratchet skip gap too large")
            if len(self.skipped_keys) + gap > MAX_SKIPPED_KEYS:
                raise ValueError("ratchet skipped key limit exceeded")
            while self.recv_n < message_n:
                mk, self.recv_chain_key = _kdf_ck(self.recv_chain_key)
                self._store_skip(peer_public, self.recv_n, mk)
                self.recv_n += 1
            message_key, self.recv_chain_key = _kdf_ck(self.recv_chain_key)
            self.recv_n = message_n + 1

        aad = RATCHET_MAGIC + peer_public + struct.pack(">II", prev_n, message_n)
        plaintext = xchacha_decrypt(message_key, ciphertext, associated_data=aad)
        return plaintext

File: rust/yakr-core/src/ratchet.rs
//! X25519 double ratchet matching Python `yakr_core.ratchet`.

use std::collections::BTreeMap;

use base64::{engine::general_purpose::URL_SAFE, Engine as _};
use serde::{Deserialize, Serialize};
use x25519_dalek::StaticSecret;
use yakr_crypto::{hkdf_sha256, xchacha_decrypt, xchacha_encrypt};

pub const RATCHET_MAGIC: &[u8; 5] = b"YKDR2";
const ROOT_INFO: &[u8] = b"yakr/v1.0/double-ratchet-root";
const RK_INFO: &[u8] = b"yakr/v1.0/double-ratchet-rk";
const CK_INFO: &[u8] = b"yakr/v1.0/double-ratchet-ck";
const SEND_CHAIN_INFO: &[u8] = b"yakr/v1.0/double-ratchet-send";
const RECV_CHAIN_INFO: &[u8] = b"yakr/v1.0/double-ratchet-recv";

fn hkdf64(ikm: &[u8], info: &[u8], salt: &[u8]) -> ([u8; 32], [u8; 32]) {
    let mut out = [0u8; 64];
    hkdf::Hkdf::<sha2::Sha256>::new(Some(salt), ikm)
        .expand(info, &mut out)
        .expect("hkdf64");
    let mut a = [0u8; 32];
    let mut b = [0u8; 32];
    a.copy_from_slice(&out[..32]);
    b.copy_from_slice(&out[32..]);
    (a, b)
}

fn kdf_rk(root_key: &[u8; 32], dh_output: &[u8]) -> ([u8; 32], [u8; 32]) {
    hkdf64(root_key, RK_INFO, dh_output)
}

fn kdf_ck(chain_key: &[u8; 32]) -> ([u8; 32], [u8; 32]) {
    hkdf64(chain_key, CK_INFO, b"")
}

fn shared_secret(private_key: &[u8; 32], peer_public: &[u8; 32]) -> [u8; 32] {
    yakr_crypto::x25519_shared_secret(private_key, peer_public)
}

fn b64encode(data: &[u8]) -> String {
    URL_SAFE.encode(data).trim_end_matches('=').to_string()
}

fn b64decode(value: &str) -> Vec<u8> {
    let rem = value.len() % 4;
    let padded = if rem == 0 {
        value.to_string()
    } else {
        format!("{value}{}", "=".repeat(4 - rem))
    };
    URL_SAFE.decode(padded).unwrap_or_default()
}

#[derive(Debug, Clone)]
pub struct RatchetState {
    pub root_key: [u8; 32],
    pub dh_self_private: [u8; 32],
    pub dh_self_public: [u8; 32],
    pub dh_peer_public: Option<[u8; 32]>,
    pub send_chain_key: Option<[u8; 32]>,
    pub recv_chain_key: Option<[u8; 32]>,
    pub send_n: u32,
    pub recv_n: u32,
    pub prev_send_n: u32,
    pub skipped_keys: BTreeMap<String, String>,
    pub hybrid: bool,
}

#[derive(Serialize, Deserialize)]
pub struct RatchetDict {
    pub version: u32,
    pub root_key: String,
    pub dh_self_private: String,
    pub dh_self_public: String,
    #[serde(default)]
    pub dh_peer_public: String,
    #[serde(default)]
    pub send_chain_key: String,
    #[serde(default)]
    pub recv_chain_key: String,
    #[serde(default)]
    pub send_n: u32,
    #[serde(default)]
    pub recv_n: u32,
    #[serde(default)]
    pub prev_send_n: u32,
    #[serde(default)]
    pub skipped_keys: BTreeMap<String, String>,
    #[serde(default)]
    pub hybrid: bool,
}

impl RatchetState {
    pub fn from_master(master_secret: &[u8; 32], is_initiator: bool, hybrid: bool) -> Self {
        let root_key = hkdf_sha256(master_secret, ROOT_INFO, b"");
        let mut send_chain = hkdf_sha256(&root_key, SEND_CHAIN_INFO, b"");
        let mut recv_chain = hkdf_sha256(&root_key, RECV_CHAIN_INFO, b"");
        if !is_initiator {
            std::mem::swap(&mut send_chain, &mut recv_chain);
        }
        let (dh_self_private, dh_self_public) = yakr_crypto::x25519_generate_keypair();
        Self {
            root_key,
            dh_self_private,
            dh_self_public,
            dh_peer_public: None,
            send_chain_key: Some(send_chain),
            recv_chain_key: Some(recv_chain),
            send_n: 0,
            recv_n: 0,
            prev_send_n: 0,
            skipped_keys: BTreeMap::new(),
            hybrid,
        }
    }

    pub fn to_dict(&self) -> RatchetDict {
        RatchetDict {
            version: 2,
            root_key: b64encode(&self.root_key),
            dh_self_private: b64encode(&self.dh_self_private),
            dh_self_public: b64encode(&self.dh_self_public),
            dh_peer_public: self
                .dh_peer_public
                .map(|k| b64encode(&k))
                .unwrap_or_default(),
            send_chain_key: self
                .send_chain_key
                .map(|k| b64encode(&k))
                .unwrap_or_default(),
            recv_chain_key: self
                .recv_chain_key
                .map(|k| b64encode(&k))
                .unwrap_or_default(),
            send_n: self.send_n,
            recv_n: self.recv_n,
            prev_send_n: self.prev_send_n,
            skipped_keys: self.skipped_keys.clone(),
            hybrid: self.hybrid,
        }
    }

    pub fn from_dict(dict: &RatchetDict) -> Result<Self, String> {
        if dict.version != 2 {
            return Err("unsupported ratchet version; re-pair required".into());
        }
        let peer = b64decode(&dict.dh_peer_public);
        let send = b64decode(&dict.send_chain_key);
        let recv = b64decode(&dict.recv_chain_key);
        Ok(Self {
            root_key: b64decode(&dict.root_key).try_into().map_err(|_| "root_key")?,
            dh_self_private: b64decode(&dict.dh_self_private)
                .try_into()
                .map_err(|_| "dh_self_private")?,
            dh_self_public: b64decode(&dict.dh_self_public)
                .try_into()
                .map_err(|_| "dh_self_public")?,
            dh_peer_public: if peer.len() == 32 {
                Some(peer.try_into().unwrap())
            } else {
                None
            },
            send_chain_key: if send.len() == 32 {
                Some(send.try_into().unwrap())
            } else {
                None
            },
            recv_chain_key: if recv.len() == 32 {
                Some(recv.try_into().unwrap())
            } else {
                None
            },
            send_n: dict.send_n,
            recv_n: dict.recv_n,
            prev_send_n: dict.prev_send_n,
            skipped_keys: dict.skipped_keys.clone(),
            hybrid: dict.hybrid,
        })
    }

    fn header_aad(&self, prev_n: u32, message_n: u32) -> Vec<u8> {
        let mut out = Vec::with_capacity(5 + 32 + 8);
        out.extend_from_slice(RATCHET_MAGIC);
        out.extend_from_slice(&self.dh_self_public);
        out.extend_from_slice(&prev_n.to_be_bytes());
        out.extend_from_slice(&message_n.to_be_bytes());
        out
    }

    fn dh_ratchet(&mut self, peer_public: [u8; 32]) {
        self.dh_peer_public = Some(peer_public);
        let dh_output = shared_secret(&self.dh_self_private, &peer_public);
        let (root, recv) = kdf_rk(&self.root_key, &dh_output);
        self.root_key = root;
        self.recv_chain_key = Some(recv);
        self.recv_n = 0;

        let (new_private, new_public) = yakr_crypto::x25519_generate_keypair();
        self.dh_self_private = new_private;
        self.dh_self_public = new_public;
        let dh_output = shared_secret(&self.dh_self_private, &peer_public);
        let (root, send) = kdf_rk(&self.root_key, &dh_output);
        self.root_key = root;
        self.send_chain_key = Some(send);
        self.prev_send_n = self.send_n;
        self.send_n = 0;
    }

    fn store_skip(&mut self, peer_public: &[u8; 32], n: u32, message_key: [u8; 32]) {
        let key_id = format!("{}:{n}", hex::encode(peer_public));
        self.skipped_keys.insert(key_id, b64encode(&message_key));
    }

    fn skip_key(&self, peer_public: &[u8; 32], n: u32) -> Result<[u8; 32], ()> {
        let key_id = format!("{}:{n}", hex::encode(peer_public));
        let stored = self.skipped_keys.get(&key_id).ok_or(())?;
        let bytes = b64decode(stored);
        bytes.try_into().map_err(|_| ())
    }

    pub fn encrypt(&mut self, plaintext: &[u8]) -> Result<Vec<u8>, String> {
        let mut send_chain = self.send_chain_key.ok_or("send chain not initialized")?;
        let (message_key, next_chain) = kdf_ck(&send_chain);
        self.send_chain_key = Some(next_chain);
        let aad = self.header_aad(self.prev_send_n, self.send_n);
        let ciphertext = xchacha_encrypt(&message_key, plaintext, &aad).map_err(|e| format!("{e:?}"))?;
        let mut out = Vec::with_capacity(5 + 32 + 8 + ciphertext.len());
        out.extend_from_slice(RATCHET_MAGIC);
        out.extend_from_slice(&self.dh_self_public);
        out.extend_from_slice(&self.prev_send_n.to_be_bytes());
        out.extend_from_slice(&self.send_n.to_be_bytes());
        out.extend_from_slice(&ciphertext);
        self.send_n += 1;
        Ok(out)
    }

    pub fn decrypt(&mut self, payload: &[u8]) -> Result<Vec<u8>, String> {
        if payload.len() < 5 + 32 + 8 {
            return Err("ratchet payload too short".into());
        }
        if &payload[..5] != RATCHET_MAGIC {
            return Err("invalid ratchet header".into());
        }
        let peer_public: [u8; 32] = payload[5..37].try_into().unwrap();
        let prev_n = u32::from_be_bytes(payload[37..41].try_into().unwrap());
        let message_n = u32::from_be_bytes(payload[41..45].try_into().unwrap());
        let ciphertext = &payload[45..];

        if self.dh_peer_public.is_none() {
            self.dh_peer_public = Some(peer_public);
        } else if self.dh_peer_public != Some(peer_public) {
            self.dh_ratchet(peer_public);
        }

        let mut recv_chain = self.recv_chain_key.ok_or("recv chain not initialized")?;
        let message_key = if message_n < self.recv_n {
            self.skip_key(&peer_public, message_n)
                .map_err(|_| "ratchet message already received".to_string())?
        } else {
            while self.recv_n < message_n {
                let (mk, next) = kdf_ck(&recv_chain);
                recv_chain = next;
                self.store_skip(&peer_public, self.recv_n, mk);
                self.recv_n += 1;
            }
            let (mk, next) = kdf_ck(&recv_chain);
            recv_chain = next;
            self.recv_n = message_n + 1;
            mk
        };
        self.recv_chain_key = Some(recv_chain);

        let mut aad = Vec::with_capacity(5 + 32 + 8);
        aad.extend_from_slice(RATCHET_MAGIC);
        aad.extend_from_slice(&peer_public);
        aad.extend_from_slice(&prev_n.to_be_bytes());
        aad.extend_from_slice(&message_n.to_be_bytes());
        xchacha_decrypt(&message_key, ciphertext, &aad).map_err(|e| format!("{e:?}"))
    }
}
